vm/src/vm.rs

```diff
@@ -328,39 +329,99 @@ impl VirtualMachine {
         PyObject::new(PyModule {}, self.ctx.types.module_type.clone(), Some(dict))
     }
 
+    /// Similar to PyErr_NormalizeException in CPython
+    pub fn normalize_exception(
+        &self,
+        exc_type: PyObjectRef,
+        exc_val: PyObjectRef,
+        exc_tb: PyObjectRef,
+    ) -> PyResult<PyBaseExceptionRef> {
+        let args = || {
+            if self.is_none(&exc_val) {
+                vec![]
+            } else {
+                match_class!(match exc_val.clone() {
+                    tup @ PyTuple => tup.elements.clone(),
+                    exc @ PyBaseException => exc.args().elements.clone(),
+                    obj => vec![obj],
+                })
+            }
+        };
+        let exc_type = if let Ok(exc) = PyBaseExceptionRef::try_from_object(self, exc_type.clone())
+        {
+            Either::A(exc)
+        } else if let Ok(cls) = PyClassRef::try_from_object(self, exc_type.clone()) {
+            Either::B(cls)
+        } else {
+            return Err(self.new_type_error(format!(
+                "expected an exception instance or type, got '{}'",
+                exc_type.class().name
+            )));
+        };
+        let exc = if let Ok(exc) = PyBaseExceptionRef::try_from_object(self, exc_val.clone()) {
+            match exc_type {
+                Either::A(_) => {
+                    return Err(self.new_type_error(
+                        "instance exception may not have a separate value".to_string(),
+                    ))
+                }
+                Either::B(cls) => {
+                    if objtype::isinstance(&exc, &cls) {
+                        exc
+                    } else {
+                        self.new_exception_obj(cls, args())?
+                    }
+                }
+            }
+        } else {
+            match exc_type {
+                Either::A(exc) => exc,
+                Either::B(cls) => self.new_exception_obj(cls, args())?,
+            }
+        };
+        if let Some(tb) = TryFromObject::try_from_object(self, exc_tb)? {
+            exc.set_traceback(Some(tb));
```

Comment:
I like this API of having a `set_traceback` function. I'm only worried that it might not work out properly for custom derived user defined exception classes from `Exception`.